### ISSUE TITLE
When the user invokes `cargo new rust-foo`, name the resulting package `foo` instead

### ISSUE THREAD

**Issue by lfairy** [rust-lang/cargo#1532]
Every so often I spot packages with `rust-` or `-rs` in their name, e.g.

* https://crates.io/crates/rust-base58
* https://crates.io/crates/coroutine-rs (the maintainer renamed this one, thanks!)

These extra bits are redundant: the fact that it's on crates.io should tell us it's Rust already. So I think we should discourage names like this.

**Strawman proposal:** If the user invokes `cargo new` with a name of the form `rust-foo` or `foo-rs`, and the `--bin` option is *not* present, then the resulting `Cargo.toml` will have a `name` field of just `foo` instead.

I'm not sure if this'll require an RFC, but it's worth looking into as we get more users post-1.0.

**Comment by frewsxcv** [comment]
I'm neutral with this proposal. I personally agree that having `*-rust`, `*-rs`, etc. in package names is excessive, but:

1. In general, special casing can cause unpredictable results for the user (not too bad in this case though)
2. Some people could prefer having the prefix/suffix in their package name
3. A user already has to edit the `Cargo.toml` before `publish`ing, so they should see the name

**Comment by lfairy** [comment]
Thanks for the feedback.

I think we need some statistics before deciding on a solution. My original proposal only makes sense when almost all packages can and should be renamed this way.

Here are some alternatives:

* Just show a warning on `cargo new`, e.g.

        $ cargo new rust-foo
        Warning: Most packages do not need a `rust-` prefix. Consider naming your package `foo` instead.

  I like this solution. It gets the message across without being too intrusive.

* @reem on IRC also suggested banning the affixes outright. But I think that's a bit too extreme.

**Comment by lfairy** [comment]
I have statistics!

```
There are 1859 packages on crates.io
Of these, 65 (3.50%) have affixed names
Of these affixed packages, 38 (58.46%) can be renamed
```

So ~ 1.5% of packages are have affixed names, but cannot drop the affix because the plain name is already taken.

The remaining ~ 2% can be renamed.

(The code and data is at https://gist.github.com/lfairy/6afd0c51dcc685682cc5.)

**Comment by reem** [comment]
@lfairy you have to also adjust for packages that originally had an affix and renamed to have it removed, those probably make up a significant portion of the 2%.

**Comment by lfairy** [comment]
@reem That's fair.

One possible heuristic uses the package owner: if the affixed and unaffixed packages have the same owner, then it's probably a rename.

I'll have to call the API for that, as the owner isn't recorded in the registry repo. https://crates.io/api/v1/crates/maud/owners should work.

**Comment by lfairy** [comment]
@reem Okay, if we count the packages that have already been renamed, the number of relevant packages increases from 38 (2.04%) to 53 (2.85%).

The remaining 15 packages are:

* [lua-rs](https://crates.io/crates/lua-rs)
* [leveldb-rs](https://crates.io/crates/leveldb-rs)
* [obj-rs](https://crates.io/crates/obj-rs)
* [hashids-rs](https://crates.io/crates/hashids-rs)
* [rust-github](https://crates.io/crates/rust-github)
* [rust-crypto](https://crates.io/crates/rust-crypto)
* [neovim-rs](https://crates.io/crates/neovim-rs)
* [rust-termios](https://crates.io/crates/rust-termios)
* [linenoise-rust](https://crates.io/crates/linenoise-rust)
* [lmdb-rs](https://crates.io/crates/lmdb-rs)
* [cql-rust](https://crates.io/crates/cql-rust)
* [rust-graph](https://crates.io/crates/rust-graph)

For these packages, the owners of the unaffixed and affixed packages are disjoint.

Now 15 / 65 is 23%, so about a quarter of affixed packages do have an excuse.

On the other hand, this means that collisions are the exception not the rule. And arguably, when someone else has taken your favorite name, the solution *isn't* to add a `-rs` suffix, but come up with a different name altogether.

So... the original proposal doesn't sound too bad, tbh.

**Comment by lfairy** [comment]
Implemented in #1540
